[MultiQubitGates] Added Sections on Bra-Ket Algebra and Multi-Qubit Gates acting on Non-adjacent qubits.

jainvasu631 opened the pull request:
Added per discussion here: [#328 comment](https://github.com/microsoft/QuantumKatas/issues/328#issuecomment-643769281)

I have added two sections on Ket-Bra Algebra. The sections added include **Bra-Ket Representation** and **Bra-Ket Decomposition**. They are similar to the corresponding sections in SingleQubitGates.

I have added a section on the Usage of Multi-Qubit gates on Non-Adjacent Qubits. It follows the guidelines prescribed by @tcNickolas 

These changes go towards finishing the 2nd checkbox in the checklist [here](https://github.com/microsoft/QuantumKatas/issues/328#issue-602988038).

tcNickolas commented:
Thank you! I might not get to reviewing this as fast as I'd like to, but I promise I'm not ignoring you, just busy with the Q# contest...

jainvasu631 commented:
No worries. I am looking forward to the contest. 
tbh I started working on the Katas as a way to brush up my Quantum Computing Knowledge. Solving the Katas seems fun and reminds of the course I took in Quantum Computing my semester abroad.

ArthurConmy commented on tutorials/MultiQubitGates/MultiQubitGates.ipynb:
```diff
@@ -481,6 +481,177 @@
+    "\n",
+    "**In Matrix Form**:\n",
+    "\n",
+    "$\\text{CINOT}$ can also be represented in matrix form. It will be a $2^3$ matrix\n",
```
```suggestion
    "$\\text{CINOT}$ can also be represented in matrix form. It will be a $2^3 \\times 2^3$ matrix\n",
```

ArthurConmy commented on tutorials/MultiQubitGates/MultiQubitGates.ipynb:
```diff
@@ -481,6 +481,177 @@
+    "\n",
+    "These can be represented as applying the following gates on the 3 qubits:\n",
+    "\n",
+    "1. $SWAP \\otimes I$\n",
```
The fact that we implement this in Q# as `SWAP(q1,q2)` without referencing the third qubit, despite the fact that of course the gate $SWAP \\otimes I$ is applied to the whole system is a real source of confusion (https://codeforces.com/blog/entry/77614?#comment-641801 and thread below this). I think it's worth explicitly mentioning that when gates are applied to seemingly only some qubits in the system (e.g `SWAP(q1,q2)`), in fact we're applying a unitary matrix to the whole system. My suggested change on line 548 highlights this.

ArthurConmy commented on tutorials/MultiQubitGates/MultiQubitGates.ipynb:
```diff
@@ -481,6 +481,177 @@
+    "2. Apply the $\\text{CNOT}$ on 2nd and 3rd qubits.\n",
+    "3. Apply the $\\text{SWAP}$ gate again on the 1st and 2nd qubits.\n",
+    "\n",
+    "These can be represented as applying the following gates on the 3 qubits:\n",
```
```suggestion
    "These can be represented as applying the following gates on the 3 qubits. Note that we're applying a gate to the whole system, so we must apply $\\text{SWAP} \\otimes I$ since we are dealing with a 3-qubit system, despite for example when implementing this in Q# we will call `SWAP(q1,q2)` if `q1` and `q2` are the first two qubits in the system (without explicitly referencing `q3`). \n",
```

jainvasu631 commented:
@ArthurConmy I have added the requested changed. I edited them briefly and moved it to place which I felt was more appropriate. 
https://github.com/microsoft/QuantumKatas/blob/e4402ec79f7a9579f21ba4c21d93e33c615ad137/tutorials/MultiQubitGates/MultiQubitGates.ipynb#L583-L585

jainvasu631 commented:
@tcNickolas Could you take a look at this.
